tests/koji/test_koji_vmi.py

```diff
@@ -238,3 +239,144 @@ def test_koji_vmi_compound_product_name(fake_koji, koji_dir):
         signing_key=None,
         release=rel_obj,
     )
+
+
+@mark.parametrize("boot_mode", [None, "legacy", "uefi", "hybrid"])
+def test_coreos_assembler_image(boot_mode, fake_koji, koji_dir):
+    boot_mode_enum = BootMode(boot_mode) if boot_mode else None
+
+    archives = []
+    images = {
+        "meta.json": "json",
+        "commitmeta.json": "json",
+        "coreos-assembler-git.tar.gz": "tar",
+    }
+    for file_name, type_name in images.items():
+        archives.append(
+            {
+                "btype": "image",
+                "btype_id": 4,
+                "build_id": 123456,
+                "buildroot_id": 1234567,
+                "checksum": "dde60621880aa996c42e356a687744ef",
+                "checksum_type": 0,
+                "compression_type": None,
+                "filename": file_name,
+                "id": 234792,
+                "metadata_only": False,
+                "size": 16077,
+                "type_description": "JSON data",
+                "type_extensions": "json",
+                "type_id": 49,
+                "type_name": type_name,
+            }
+        )
+
+    name = "rhcos"
+    version = "4.11"
+    release = "1"
+    nvr = f"{name}-{version}-{release}"
+
+    source = Source.get(
+        f"koji:https://koji.example.com/?vmi_build={nvr}", basedir=koji_dir
+    )
+    assert not fake_koji.last_url
+
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": name,
+        "version": version,
+        "release": release,
+        "nvr": nvr,
+        "completion_time": "2022-12-20 12:30:25",
+        "extra": {"typeinfo": {"image": {"arch": "x86_64", "boot_mode": boot_mode}}},
+    }
+
+    fake_koji.build_data[nvr] = fake_koji.build_data[1234]
+
+    fake_koji.insert_archives(archives=archives, build_nvr=nvr)
+
+    path = "/path/to/rhcos/metadata"
+    meta_data = """{    
+    "name": "rhcos",
+    "summary": "OpenShift 4",
+    "coreos-assembler.container-config-git":{
+            "branch": "release-4.11"
+    },
+    "images": {
+        "azure": {
+            "path": "rhcos-4.11-0-azure.x86_64.vhd.gz",
+            "sha256": "2cd817331af29093e2eaa025139ebddd6008c193970b06b35afbbdbebae0ce3e",
+            "dest": ["https://example.windows.net/imagebucket/rhcos-4.11-0-azure.x86_64.vhd"]},
+  "aws": {
+        "path": "rhcos-4.11-0-aws.x86_64.vmdk.gz",
+        "sha256": "4ef7806152bd89ce44326ff746c4f1883ad543885e980bce59821df2d946ea4c"
+        }},
+ "azure": {
+        "image": "rhcos-4.11-0-azure.x86_64.vhd",
+        "url": "https://example.windows.net/imagebucket/rhcos-4.11-0-azure.x86_64.vhd"},
+ "amis": [
+    {"name": "us-gov-west-1",
+   "hvm": "ami-01"
+   },
+    {"name": "us-east-1",
+   "hvm": "ami-02"
+   }]}"""
+
+    with patch("builtins.open", mock_open(read_data=meta_data)) as mocked_open:
```

Comment:
What I wrote in comment https://github.com/release-engineering/pushsource/pull/467#discussion_r1508205061 applies here too.  Please, don't mock `open`, it's completely unnecessary and it makes the test fragile as it now makes the test depend on the code-under-test calling `open` exactly once and only for this purpose.  If someone needs to add code later which opens something else, they will be forced to rewrite this test.

There should be no reason to mock any `open` or `path` methods here.  Please, check the value of the `koji_dir` variable while this test runs.  It ought to be a directory that you can write to.  Just arrange for the JSON file to really exist under that directory and then let the code under test really read it from that directory.  Adding mocks unnecessarily makes the tests more complicated and fragile.